symbolic-debuginfo/src/sourcebundle/utf8_reader.rs

```diff
@@ -0,0 +1,310 @@
+
+use std::io::{BufRead, Error, ErrorKind, Read, Result};
+use thiserror::Error;
+
+const MAX_UTF8_SEQUENCE_SIZE: usize = 4;
+const MAX_BUFFER_SIZE: usize = 8 * 1024;
+
+#[derive(Debug, Error)]
+#[error("Invalid UTF-8 sequence")]
+pub struct UTF8ReaderError {
+    /// Make it impossible to construct this struct outside of this module.
+    _a: (),
+}
+
+impl UTF8ReaderError {
+    fn new() -> Self {
+        Self { _a: () }
+    }
+}
+
+pub struct Utf8Reader<R> {
+    inner: R,
+
+    /// buffer_vec always contains a valid UTF-8 sequence. It is possible that
+    /// buffer_vec[buffer_pos..] does not contain a valid UTF-8 sequence, if
+    /// buffer_pos is in the middle of a multi-byte UTF-8 character.
+    buffer_vec: Vec<u8>,
+    buffer_pos: usize,
+}
+
+impl<R> Utf8Reader<R> {
+    pub fn new(inner: R) -> Self {
+        Self {
+            inner,
+            buffer_vec: vec![],
+            buffer_pos: 0,
+        }
+    }
+
+    fn buffer(&self) -> &[u8] {
+        &self.buffer_vec[self.buffer_pos..]
+    }
+}
+
+impl<R> Utf8Reader<R>
+where
+    R: Read,
+{
+    /// Reads a UTF-8 sequence from the inner reader into the buffer, returning the number
+    /// of bytes read. Errors if this is not possible.
+    ///
+    /// The function guarantees that the function will return an Ok variant with a positive
+    /// number of bytes read if it is possible to read a valid UTF-8 sequence from the inner
+    /// reader. The function also guarantees that all of the bytes read from the inner reader will
+    /// be stored in in the buffer, at indices up to the value returned by the function (in other
+    /// words, no bytes are lost).
+    ///
+    /// Panics if the buffer is not at least `MAX_UTF8_SEQUENCE_SIZE` bytes long.
+    fn read_utf8(&mut self, buf: &mut [u8]) -> Result<usize> {
+        if buf.len() < MAX_UTF8_SEQUENCE_SIZE {
+            panic!("Buffer needs to be at least {MAX_UTF8_SEQUENCE_SIZE} bytes long");
+        }
+
+        // We need to leave at least three bytes in the buffer in case the first read
+        // ends in the middle of a UTF-8 character. In the worst case, we end at the first
+        // byte of a 4-byte UTF-8 character, and we need to read 3 more bytes to get a
+        // valid sequence.
+        let bytes_to_read = buf.len() - MAX_UTF8_SEQUENCE_SIZE + 1;
+        let read_buf = &mut buf[..bytes_to_read];
+
+        let bytes_read = self.inner.read(read_buf)?;
+        let read_portion = &buf[..bytes_read];
+
+        let valid_up_to = utf8_up_to(read_portion);
+        let invalid_portion = &read_portion[valid_up_to..];
+
+        if invalid_portion.len() >= MAX_UTF8_SEQUENCE_SIZE {
+            return Err(Error::new(ErrorKind::InvalidData, UTF8ReaderError::new()));
+        }
+
+        // We use this buffer to read up to 4 bytes from the inner reader
+        // to try to get a valid UTF-8 sequence.
+        let mut next_character_buffer = Vec::from(invalid_portion);
```

Comment:
You're allocating this vector on every call. Might be worth it to keep it as a struct member and reuse it.

Reply:
I figured out how to read it directly into the buffer, no vector needed